src/jdk.incubator.code/share/classes/jdk/incubator/code/type/CoreTypeFactory.java

```diff
@@ -118,27 +119,82 @@ public JavaType constructType(TypeElement.ExternalizedTypeElement tree) {
                         constructTypeArgument(tree, 0, NO_WILDCARDS));
             } else if (identifier.startsWith("#")) {
                 // type-var
-                if (tree.arguments().size() != 1) {
+                if (tree.arguments().size() != 2) {
                     throw badType(tree, "type variable");
                 }
-                String[] parts = identifier.substring(1).split("::");
-                if (parts.length == 2) {
-                    // class type-var
-                    return JavaType.typeVarRef(parts[1],
-                            (ClassType)constructType(parseExTypeElem(parts[0])),
-                            constructTypeArgument(tree, 0, NO_WILDCARDS));
-                } else if (parts.length == 3) {
-                    // method or constructor type-var
-                    String desc = String.format("%s::%s", parts[0], parts[1]);
-                    TypeVarRef.Owner owner = parts[1].startsWith("<new>") ?
-                            parseConstructorRef(desc) :
-                            parseMethodRef(String.format("%s::%s", parts[0], parts[1]));
-                            return JavaType.typeVarRef(parts[2],
-                                    owner,
-                                    constructTypeArgument(tree, 0, NO_WILDCARDS));
-                } else {
+
+                String name = tree.identifier().substring(1);
+                if (name.isEmpty()) {
+                    throw badType(tree, "type variable");
+                }
+
+                if (!(constructType(tree.arguments().get(0)) instanceof TypeVarRef.Owner owner)) {
                     throw badType(tree, "type variable");
                 }
+
+                return JavaType.typeVarRef(name,
```

Comment:
This is much nicer! No longer need to monkey around with string representation -- and can rely on nested type elements instead!